### PR TITLE
PoA_DynamicFederationMembers_3

### PR DESCRIPTION
**Pull request by noescape00**
Refactoring, comments, tests, added partial implementation of voting manager.

### CODE REVIEW

File: src/Stratis.Bitcoin.Features.PoA.Tests/VotingManagerTests.cs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+namespace Stratis.Bitcoin.Features.PoA.Tests
+{
+    public class VotingManagerTests : PoATestsBase
+    {
+        private readonly VotingManager votingManager;
+
+        public VotingManagerTests()
+        {
+            string dir = TestBase.CreateTestDir(this);
+            var keyValueRepo = new KeyValueRepository(dir, new DBreezeSerializer(this.network));
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.PoA.Tests/VotingManagerTests.cs]: This are not integration tests I wonder if you should not mock the KeyValue repo?

Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA.Tests/VotingManagerTests.cs]: No, why? 
If I can use actual implementation then it's always better to do so. 
Things will be saved to temp folder anyway.

File: src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        {
+            lock (this.locker)
+            {
+                return new List<Poll>(this.finishedPolls);
+            }
+        }
+
+        // TODO subscribe to signals
+        // TODO more logs and log polls
+        private void onBlockConnected(ChainedHeaderBlock chBlock)
```
Comment (dangershony) [inline src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs]: Come to think about it, I am not sure the voting db should be disconnected from consensus, it should probably be a full validation rule  that will update some db and that db will also rewind together with the engine rewind.

File: src/Stratis.Bitcoin.Features.PoA.Tests/VotingManagerTests.cs (hunk 1 above)
Comment (dangershony) [reply src/Stratis.Bitcoin.Features.PoA.Tests/VotingManagerTests.cs]: Yeah but there is no added benefit in this case you are testing `KeyValueRepository` which  has its own tests, using a mock will skip db and made the subject being tested much faster.

Note this is for unit tests I refer to no integration tests.

Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA.Tests/VotingManagerTests.cs]: When actual implementation is used we are testing it indirectly which makes all tests more robust.

File: src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs (hunk 2 above)
Comment (noescape00) [reply src/Stratis.Bitcoin.Features.PoA/Voting/VotingManager.cs]: As discussed in PM this is what happens- on block connect and disconnect are called under a reorg lock after block was connected and rewind happened. So due to the lack of other place to do this- this is the only place to add this logic.

**Pull request merged**